### PR TITLE
Custom attribute limits

### PR DESCRIPTION
**Pull request by gingerbenw**
## Goal

Add validation for custom span attributes

## Changeset

- Add new configuration options for attribute validation
- Add new validation method for truncating string and array attributes based on configuration options
- Add new validation method for dropping spans based on configuration options
- Add logging for modified spans

## Testing

Add unit tests for all attribute limits and behaviours

### CODE REVIEW

File: packages/core/lib/attributes.ts [hunk 1]
```diff
@@ -34,15 +35,64 @@ export interface SpanAttributesSource <C extends Configuration> {
...
+      const truncatedValue = value.slice(0, this.spanAttributeLimits.attributeArrayLengthLimit)
+      this.logger.warn('Attribute array length limit reached. Discarding excess array items.')
+      this.set(name, truncatedValue)
+    }
   }
 
   set (name: string, value: SpanAttribute) {
     if (typeof name === 'string' && (typeof value === 'string' || typeof value === 'boolean' || isNumber(value) || Array.isArray(value))) {
+      if (!this.attributes.has(name) && this.attributes.size >= this.spanAttributeLimits.attributeCountLimit) {
+        this.logger.warn('Attribute count limit reached. Discarding attribute.')
```
Comment (gingerbenw) [inline packages/core/lib/attributes.ts:91]: I would be keen to update this message wording to something like `Dropping attribute with key "really-long-key". Key length exceeds limit of 128 characters.`

File: packages/core/lib/attributes.ts [hunk 2]
```diff
@@ -34,15 +35,64 @@ export interface SpanAttributesSource <C extends Configuration> {
...
+      return
+    }
+
+    if (typeof value === 'string' && value.length > this.spanAttributeLimits.attributeStringValueLimit) {
+      this.attributes.set(name, truncateString(value, this.spanAttributeLimits.attributeStringValueLimit))
+    }
+
+    if (Array.isArray(value) && value.length > this.spanAttributeLimits.attributeArrayLengthLimit) {
+      const truncatedValue = value.slice(0, this.spanAttributeLimits.attributeArrayLengthLimit)
+      this.logger.warn('Attribute array length limit reached. Discarding excess array items.')
```
Comment (gingerbenw) [inline packages/core/lib/attributes.ts:83]: `Attribute array length limit of 10 reached. Discarding the last 5 array items.` ?

File: packages/core/lib/attributes.ts [hunk 3]
```diff
@@ -34,15 +35,64 @@ export interface SpanAttributesSource <C extends Configuration> {
...
+  constructor (initialValues: Map<string, SpanAttribute>, spanAttributeLimits: SpanAttributesLimits, logger: Logger) {
     this.attributes = initialValues
+    this.spanAttributeLimits = spanAttributeLimits
+    this.logger = logger
+  }
+
+  // Validate after the span has ended and after the sampling
+  private validateAttribute (name: string, value: SpanAttribute) {
+    if (name.length > ATTRIBUTE_KEY_LENGTH_LIMIT) {
+      this.logger.warn('Attribute key limit reached. Discarding attribute.')
```
Comment (gingerbenw) [inline packages/core/lib/attributes.ts:71]: I would be keen to update this message wording to something like Dropping attribute with key "really-long-key". Key length exceeds limit of 128 characters.

File: packages/core/lib/attributes.ts (hunk 1 above)
Comment (gingerbenw) [inline packages/core/lib/attributes.ts:91]: Suggestions for improved message wording: `Attribute count limit (64) reached. Discarding attribute "attribute-name".`

File: packages/core/lib/config.ts [hunk 4]
```diff
@@ -127,6 +138,21 @@ export const schema: CoreSchema = {
     defaultValue: undefined,
     message: 'should be an array of functions',
     validate: isOnSpanEndCallbacks
+  },
+  attributeStringValueLimit: {
+    defaultValue: ATTRIBUTE_STRING_VALUE_LIMIT_DEFAULT,
+    message: 'should be a number',
```
Comment (yousif-bugsnag) [inline packages/core/lib/config.ts:144]:
```suggestion
    message: `should be a number between 1 and ${ATTRIBUTE_STRING_VALUE_LIMIT_MAX}`,
```

File: packages/core/lib/config.ts [hunk 5]
```diff
@@ -127,6 +138,21 @@ export const schema: CoreSchema = {
...
+    validate: (value: unknown): value is number => isNumber(value) && value > 0 && value <= ATTRIBUTE_STRING_VALUE_LIMIT_MAX
+  },
+  attributeArrayLengthLimit: {
+    defaultValue: ATTRIBUTE_ARRAY_LENGTH_LIMIT_DEFAULT,
+    message: 'should be a number',
```
Comment (yousif-bugsnag) [inline packages/core/lib/config.ts:149]:
```suggestion
    message: `should be a number between 1 and ${ATTRIBUTE_STRING_VALUE_LIMIT_MAX}`,
```

File: packages/core/lib/config.ts [hunk 6]
```diff
@@ -127,6 +138,21 @@ export const schema: CoreSchema = {
...
+    validate: (value: unknown): value is number => isNumber(value) && value > 0 && value <= ATTRIBUTE_ARRAY_LENGTH_LIMIT_MAX
+  },
+  attributeCountLimit: {
+    defaultValue: ATTRIBUTE_COUNT_LIMIT_DEFAULT,
+    message: 'should be a number',
```
Comment (yousif-bugsnag) [inline packages/core/lib/config.ts:154]:
```suggestion
    message: `should be a number between 1 and ${ATTRIBUTE_COUNT_LIMIT_MAX}` ,
```

File: packages/core/lib/span.ts [hunk 7]
```diff
@@ -64,6 +64,17 @@ export function spanToJson (span: SpanEnded, clock: Clock): DeliverySpan {
   }
 }
 
+export function spanEndedToSpan (span: SpanEnded): Span {
+  return {
+    id: span.id,
+    traceId: span.traceId,
+    samplingRate: span.samplingRate,
```
Comment (yousif-bugsnag) [inline packages/core/lib/span.ts:71]: Possibly beyond the scope of this PR, but it feels like these properties should be made immutable in the same way as in `SpanFactory.toPublicApi()`

Also, should we expose the span name here and in the public Span API (also as an immutable property)? If you're writing an onSpanEnd callback it seems like that would be useful to know.

Comment (lemnik) [reply packages/core/lib/span.ts:71]: Yeah. `name` should definitely be public (and immutable)

**Review by tomlongridge — COMMENTED** [review]
LGTM

**Review by lemnik — APPROVED** [review]

**Review by yousif-bugsnag — APPROVED** [review]

**Pull request merged**